src/main/java/cloud/prefab/client/config/ConfigResolver.java

```diff
@@ -1,221 +1,371 @@
+  private static final Logger LOG = LoggerFactory.getLogger(ConfigResolver.class);
 
-  private final PrefabCloudClient baseClient;
-  private final ConfigLoader configLoader;
-  private final AtomicReference<ImmutableMap<String, ResolverElement>> localMap = new AtomicReference<>(
-    ImmutableMap.of()
-  );
+  private final ConfigStore configStore;
+  private final WeightedValueEvaluator weightedValueEvaluator;
 
   private long projectEnvId = 0;
 
-  public ConfigResolver(PrefabCloudClient baseClient, ConfigLoader configLoader) {
-    this.baseClient = baseClient;
-    this.configLoader = configLoader;
+  public ConfigResolver(ConfigStore configStoreImpl) {
+    this(configStoreImpl, 0L);
+  }
+
+  public ConfigResolver(ConfigStore configStoreImpl, long projectEnvId) {
+    this.weightedValueEvaluator = new WeightedValueEvaluator();
+    this.projectEnvId = projectEnvId;
+    this.configStore = configStoreImpl;
   }
 
   public Optional<Prefab.ConfigValue> getConfigValue(String key) {
-    final ResolverElement resolverElement = localMap.get().get(key);
-    if (resolverElement != null) {
-      return Optional.of(resolverElement.getConfigValue());
-    }
-    return Optional.empty();
+    return getConfigValue(key, new HashMap<>());
   }
 
-  public Optional<Prefab.Config> getConfig(String key) {
-    final ResolverElement resolverElement = localMap.get().get(key);
-    if (resolverElement != null) {
-      return Optional.of(resolverElement.getConfig());
+  public Optional<Prefab.ConfigValue> getConfigValue(
+    String key,
+    Map<String, Prefab.ConfigValue> properties
+  ) {
+    if (!configStore.containsKey(key)) {
+      return Optional.empty();
+    }
+    final ConfigElement configElement = configStore.getElement(key);
+
+    final Optional<Match> match = findMatch(configElement, properties);
+
+    if (match.isPresent()) {
+      return Optional.of(match.get().getConfigValue());
+    } else {
+      return Optional.empty();
     }
-    return Optional.empty();
   }
 
   /**
-   * Return the changed config values since last update()
+   * find if we have a match for the given properties
+   *
+   * @param configElement
+   * @param properties
+   * @return
    */
-  public synchronized List<ConfigChangeEvent> update() {
-    // store the old map
-    final Map<String, Prefab.ConfigValue> before = localMap
-      .get()
-      .entrySet()
-      .stream()
-      .collect(Collectors.toMap(Entry::getKey, e -> e.getValue().getConfigValue()));
+  Optional<Match> findMatch(
+    ConfigElement configElement,
+    Map<String, Prefab.ConfigValue> properties
+  ) {
+    // Prefer rows that have a projEnvId to ones that don't
+    // There will be 0-1 rows with projenv and 0-1 rows without (the default row)
+    final Optional<Match> match = configElement
+      .getRowsProjEnvFirst(projectEnvId)
+      .map(configRow -> {
+        Map<String, Prefab.ConfigValue> rowProperties = new HashMap<>(properties);
```

Comment:
its prudent to construct a HashMap with the right initial capacity if its known - here i think it is because thats properties.size() +configRow.getPropertiesMap().size()

Reply:
The HashMap constructor that takes an initial capacity is a bit of a troll because the default load factor is 0.75. So essentially you're guaranteeing that the map will be undersized and need to grow (and when it grows it will be oversized). Guava has a method that does the math for you:
https://github.com/google/guava/blob/553037486901cc60820ab7dcb38a25b6f34eba43/guava/src/com/google/common/collect/Maps.java#L272-L302

But if we've already taken a Guava dep it seems like we should probably just use Guava's immutable collections across the board?

Reply:
Good point, I forget which direction the default load factor forces things in

Reply:
:trollface: